### PR TITLE
Feat/add visited country

### PR DESCRIPTION
**Pull request by NickKlock**

### CODE REVIEW

File: backend/src/main/java/de/neuefische/backend/service/UserService.java [hunk 1]
```diff
@@ -37,4 +38,18 @@ public User findUserById(String id){
                 .orElseThrow(() -> new NoSuchElementException("User with id " + id + "does not exsist"));
     }
 
+    public User updateUser(String id, User editedUser){
+        checkIfExsists(id);
+        return userRepo.save(editedUser);
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/neuefische/backend/service/UserService.java:43]: How do you make sure the parameter id is the same as editedUser.id ?

File: backend/src/main/java/de/neuefische/backend/service/UserService.java [hunk 2]
```diff
@@ -37,4 +38,18 @@ public User findUserById(String id){
...
+    }
+
+    private void checkIfExsists(String id) throws NoSuchUserException {
+        for (User user :
+                listUsers()) {
+            if (user.id().equals(id)) {
+                return;
+            }
+        }
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/neuefische/backend/service/UserService.java:52]: Load all user just to check if one exists sound pretty expensive...
I would use the appropriate mongodb method.

File: backend/src/test/java/de/neuefische/backend/controller/RestCountriesControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+@AutoConfigureMockMvc
+class RestCountriesControllerTest {
+    String endPoint = "/api/countries";
+
+    @Autowired
+    private MockMvc mvc;
+
+
+    @Test
+    void get_countries_returns_a_list_of_countries_and_deserialize()
```
Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/backend/controller/RestCountriesControllerTest.java:31]: the convention is <method-name>_when<initial-values>_then<expected-outcome>

```suggestion
    void getCountries_returnsAListOfCountriesAndDeserialize()
```

Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/backend/controller/RestCountriesControllerTest.java:31]: This looks pretty much like an unit test. you're only testing RestCountriesWebClient class, so i would extract this test to its own file

File: backend/src/test/java/de/neuefische/backend/controller/RestCountriesControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+        RestCountriesWebClient restCountriesWebClient =
+                new RestCountriesWebClient(mockWebServer.url("/").url().toString());
+
+        mockWebServer.enqueue(
+                new MockResponse().setResponseCode(200)
+                        .setHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
+                        .setBody(givenJson));
+
+        restCountriesWebClient.sendGetRequest();
```
Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/backend/controller/RestCountriesControllerTest.java:94]: Why do you not assert that the result is what you expect?

File: backend/src/test/java/de/neuefische/backend/controller/RestCountriesControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+        assertThat(request.getMethod()).isEqualTo("GET");
+
+        mockWebServer.close();
+    }
+
+    @Test
+    void getCountries_expect_status_ok() throws Exception {
+        mvc.perform(get(endPoint))
+                .andExpect(status().isOk());
```
Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/backend/controller/RestCountriesControllerTest.java:105]: Why dont you mock the webserver here as well?
And why dont you assert the response?

File: backend/src/test/java/de/neuefische/backend/controller/UserControllerTest.java [hunk 6]
```diff
@@ -53,9 +54,7 @@ void addUser_expect_valid_user_and_an_id() throws Exception {
 
         User actualResult = objectMapper.readValue(result.getResponse().getContentAsString(), User.class);
 
-        assertInstanceOf(User.class, actualResult);
         assertFalse(actualResult.id().isEmpty());
-        assertFalse(actualResult.id().isBlank());
```
Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/backend/controller/UserControllerTest.java:58]: You should check if the name and visitedCountries got save like you expected it.

File: backend/src/test/java/de/neuefische/backend/service/RestCountriesServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class RestCountriesServiceTest {
+
+    RestCountriesService restCountriesService = new RestCountriesService();
+
+    @Test
+    void refactorApiResponse_deseralize_properly() throws JsonProcessingException, RestCountryException {
```
Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/backend/service/RestCountriesServiceTest.java:19]: You are only testing the mapping from an array to a list and not the deserialization this happens in your other class within the webclient

File: backend/src/test/java/de/neuefische/backend/service/RestCountriesServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                    }
+                    ]
+                """;
+        ObjectMapper objectMapper = new ObjectMapper();
+        RestCountry[] restCountries = objectMapper.readValue(givenJson, RestCountry[].class);
+
+        List<Country> countries = restCountriesService.refactorApiResponse(restCountries);
+        assertThat(countries.get(0).name()).isEqualTo("Mauritania");
+        assertThat(countries.get(0).flag()).isEqualTo("🇲🇷");
+        assertThat(countries.get(0).threeLetterCode()).isEqualTo("MRT");
```
Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/backend/service/RestCountriesServiceTest.java:76]: Why do you only check for Mauritania? And not for the whole list? There is a second element aswell

File: backend/src/test/java/de/neuefische/backend/service/UserServiceTest.java [hunk 9]
```diff
@@ -50,17 +55,38 @@ void deleteUser_expect_correct_id() {
...
         //when
         when(userRepo.findById(givenUser.id())).thenReturn(Optional.of(givenUser));
         User userResult = userService.findUserById(givenUser.id());
         //then
-        assertEquals(givenUser,userResult);
+        assertEquals(givenUser, userResult);
+    }
+
+    @Test
+    void updateUser_expect_exception() {
```
Comment (Flooooooooooorian) [inline backend/src/test/java/de/neuefische/backend/service/UserServiceTest.java:73]: These names or not very meaning full, maybe you could add why you are expecting an exception and why not

File: frontend/src/component/AddVisitCountry.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+            .catch(console.error)
+    }
+
+    return (
+        <div>
+            <div>
+                <input type={"search"} onChange={handleSearchText} placeholder={"Search for a country.."}/>
+            </div>
+            <div className={"card-display"}>
+                {filteredCountries.length < 5 && filteredCountries.map((value) => {
```
Comment (Flooooooooooorian) [inline frontend/src/component/AddVisitCountry.tsx:48]: Why are you only displaying your list of countries when there a less than 5? 🤔

File: frontend/src/component/UserCard.tsx [hunk 11]
```diff
@@ -21,9 +21,21 @@ export default function UserCard(props: UserCardProps) {
...
+    function handleAddCountry() {
+        navigate("/overview/"+props.user.id+"/countries")
+    }
+
     return (<div>
             <h2>Name:{props.user.name} </h2>
-            {props.loggedInUser.id === props.user.id && <button onClick={handleDeleteUser}>delete User</button>}
+            <div>{props.user.visitedCountries &&
+                props.user.visitedCountries.map((c) =>
+                {return <p key={c.threeLetterCode}> {c.name}[{c.threeLetterCode}] {c.flag}</p>})}
```
Comment (Flooooooooooorian) [inline frontend/src/component/UserCard.tsx:32]:
```suggestion
                props.user.visitedCountries.map((country) => <p key={country.threeLetterCode}> {country.name}[{country.threeLetterCode}] {country.flag}</p>})
```

File: frontend/src/component/UserOverview.tsx [hunk 12]
```diff
@@ -6,9 +6,11 @@ type UserOverviewProps = {
     loggedInUser:User
     setLoggedInUser(user: User): void;
 }
+
 export default function UserOverview(props:UserOverviewProps) {
     const {users, deleteUserByID} = useUsers()
 
+    console.log(users)
```
Comment (Flooooooooooorian) [inline frontend/src/component/UserOverview.tsx:13]:
```suggestion
```

**Review by Flooooooooooorian — COMMENTED** [review]
Next time split this PR in the updateUser Part and the external API part.

Left you some comments with a few changes but overall pretty good 👍

File: backend/src/main/java/de/neuefische/backend/RestCountriesWebClient.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    public RestCountriesWebClient(String baseUrl) {
+        this.baseUrl = baseUrl;
+        setUp();
+    }
+
+    private void setUp(){
+        final int size = 16 * 1024 * 1024;
+        final ExchangeStrategies strategies = ExchangeStrategies.builder()
+                .codecs(codecs -> codecs.defaultCodecs().maxInMemorySize(size))
+                .build();
```
[the remark this answers is not in the document]
Comment (NickKlock) [reply backend/src/main/java/de/neuefische/backend/RestCountriesWebClient.java:22]: The response from the api is pretty big and we ran into an error where the memory wasn't enough. So we had to change the memory size

File: backend/src/test/java/de/neuefische/backend/service/RestCountriesServiceTest.java (hunk 8 above)
Comment (NickKlock) [reply backend/src/test/java/de/neuefische/backend/service/RestCountriesServiceTest.java:76]: when the first is parsed properly, the second one should be too.

File: frontend/src/component/AddVisitCountry.tsx (hunk 10 above)
Comment (NickKlock) [reply frontend/src/component/AddVisitCountry.tsx:48]: Otherwise its very Hugh and looks ugly

Comment (Flooooooooooorian) [reply frontend/src/component/AddVisitCountry.tsx:48]: Maybe you could only display the first 5 even if there are more?

File: backend/src/test/java/de/neuefische/backend/service/RestCountriesServiceTest.java (hunk 8 above)
Comment (Flooooooooooorian) [reply backend/src/test/java/de/neuefische/backend/service/RestCountriesServiceTest.java:77]: Yeah you should be right, but i would make sure im right. And its not that much work

**Review by Flooooooooooorian — APPROVED** [review]

**Pull request merged**